doc/cprover-manual/contracts-assigns.md

```diff
@@ -1,75 +1,368 @@
+An lvalue target `expr` with a complete type `expr_t` specifies that the range
+starting at `&expr` and of size `sizeof(expr_t)` bytes is assignable.
+
+Lvalue expressions can also be wrapped in `__CPROVER_typed_target` with the same
+meaning: for an lvalue `expr` with a complete type `expr_t`,
+`__CPROVER_typed_target(expr)` specifies that the range of `sizeof(expr_t)`
+bytes starting at `&expr` is assignable:
 
-Given a pointer `ptr` pointing into some object `o`, `__CPROVER_object_from(ptr)` 
-specifies that all bytes starting from the given pointer and until the end of 
-the object are assignable:
 ```c
-__CPROVER_size_t __CPROVER_object_from(void *ptr); 
+void __CPROVER_typed_target(expr_t expr);
 ```
 
-Given a pointer `ptr` pointing into some object `o`, `__CPROVER_object_from(ptr, size)` 
-specifies that `size` bytes starting from the given pointer and until the end of the object are assignable.
-The `size` value must such that `size <= __CPROVER_object_size(ptr) - __CPROVER_pointer_offset(ptr)` holds:
+In order to specify that a memory location which contents is interpreted as a
+pointer by the program is assignable, one must use the notation
+`__CPROVER_assigns(ptr)` or the equivalent
+`__CPROVER_assigns(__CPROVER_typed_target(ptr))`, as opposed to the slice
+operators `__CPROVER_object_whole`, `__CPROVER_object_from`,
+or `__CPROVER_object_upto`.
+This ensures that during call-by-contract replacement the memory location get
+turned into non-deterministic pointers.
+
+For instance:
 
 ```c
-__CPROVER_size_t __CPROVER_object_slice(void *ptr, __CPROVER_size_t size);
+struct circular_buffer_t {
+  int *first;
+  int *last;
+  int *current;
+}
+
+void step(struct circular_buffer_t *buf)
+// correct
+__CPROVER_assigns(__CPROVER_typed_target(buf->current))
+// not correct
+__CPROVER_assigns(__CPROVER_object_upto(&(buf->current), sizeof(buf->current))
+{
+  if(buf->current == buf->last)
+    buf->current = buf->first;
+  else
+    buf->current += 1;
+}
 ```
 
-Caveats and limitations: The slices in question must *not*
-be interpreted as pointers by the program. During call-by-contract replacement, 
-`__CPROVER_havoc_slice(ptr, size)` is used to havoc these targets, 
-and `__CPROVER_havoc_slice` does not support havocing pointers. 
-### Parameters
+### Object slice targets
+
+```c
+void __CPROVER_object_upto(void *ptr, __CPROVER_size_t size);
+```
+
+Given a pointer `ptr` pointing into some object (possibly at some non-zero offset),
+`__CPROVER_object_upto(ptr, size)` specifies that the range of `size` bytes starting
+at `ptr` is assignable.
 
-An _assigns_ clause currently supports simple variable types and their pointers,
-structs, and arrays.  Recursive pointer structures are left to future work, as
-their support would require changes to CBMC's memory model.
+
+The value of `size` must such that the range does not exceed
+the object's boundary, i.e.
+`size <= __CPROVER_OBJECT_SIZE(ptr) - __CPROVER_POINTER_OFFSET(ptr)` must hold.
```

Comment:
Just curious: what happens when the user violates this rule?

Reply:
and assertion will be violated

Reply:
added precision